Add post (#23)
orhun/personal-blog: content/automated-rust-releases.md

- [`cargo-unleash`](https://github.com/paritytech/cargo-unleash): release automatisation tooling for massiv mono-repos.

We can list other tools here, but in a nutshell, all we want to do is:

1. Make sure the project is in a _state for the release_. This might include updated version/changelog/dependencies.
2. Tag the new release.
3. Publish the new version on the various platforms (crates.io, GitHub, etc.)

**Q**: Okay okay, but aren't you overcomplicating things? I mean... just update the version in `Cargo.toml` and run `cargo publish`.

**A**: That's fair, but how about binary releases? As someone who has spent a good amount of time automating NPM releases, I would say this might not be _that easy_ for every project.

**Q**: I still think you can just copy/paste a GitHub Actions workflow file from another project and adopt it.

**A**: It is still too much manual work + release process is still not fully automated - you need to create tags/update changelog etc. Besides, your name is "**Q**" but you are not really asking questions anymore and just rambling. Maybe it is time to introduce you as a new character in these blog posts?

> **Q**: It do be like that sometimes.

Author: meme :D

https://knowyourmeme.com/memes/it-really-do-be-like-that-sometimes

> [**git-cliff**](https://github.com/orhun/git-cliff) is a command-line tool (written in [Rust](https://www.rust-lang.org/)) that provides a highly customizable way to generate changelogs from git history. It supports using [custom regular expressions](/docs/configuration#commit_parsers) to alter changelogs which are mostly based on [conventional commits](/docs/configuration#conventional_commits). With a single [configuration file](/docs/configuration), a big variety of formats can be applied for a changelog, thanks to the Jinja2/Django-inspired [template engine](/docs/category/templating). More information and examples can be found in the [GitHub repository](https://github.com/orhun/git-cliff).

For our changelog format, we can simply go with the default configuration which outputs something like the following:

```sh
# Changelog

All notable changes to this project will be documented in this file.

## [unreleased]

### Documentation

- Add link to emacs package support git-cliff (#307)
```

For this configuration, first we should install `git-cliff`:

```sh

> cargo install git-cliff

Reviewer: ```suggestion
cargo install git-cliff --locked
```
more secure!

## [unreleased]

### 📚 Documentation

- _(readme)_ Add link to emacs package support git-cliff ([#307](https://github.com/orhun/git-cliff/issues/307)) - ([fa471c7](https://github.com/orhun/git-cliff/commit/fa471c7178dce184ca6fe5bbb24b9c2db96d68ce))
```

Great, now we are ready to set up the actual releaser tool!

---

### **release-plz**

> [release-plz](https://release-plz.ieni.dev) creates a fully automated release pipeline, allowing you to easily release changes more frequently, without the fear of doing typos or other subtle manual mistakes you can make when releasing from your terminal.

Unlike `git-cliff`, we don't actually need to install `release-plz` as a command-line tool. It is because `release-plz` runs from the CI (GitHub Actions) and we are not doing anything locally for the releases. However, we still need to tweak some things regarding how we want to release our project.

For configuring `release-plz`, create `release-plz.toml` with the following contents at the root of your repository:

```toml
[workspace]
# path of the git-cliff configuration
changelog_config = "cliff.toml"

# enable changelog updates

> changelog_update = true

Reviewer: `changelog_update`, `git_tag_enable`, `allow_dirty`, `publish_allow_dirty` are the default values. Why do you override them? 🤔

Author: Just to show the reader that they are available options :)

# update dependencies with `cargo update`
dependencies_update = true

# create tags for the releases
git_tag_enable = true

# disable GitHub releases
git_release_enable = false

# labels for the release PR
pr_labels = ["release"]

# disallow updating repositories with uncommitted changes
allow_dirty = false

# disallow packaging with uncommitted changes
publish_allow_dirty = false

# disable running `cargo-semver-checks`

> semver_check = false

Reviewer: why do you disable cargo-semver-checks?
It is automatically disabled for binaries.

```

The important options here are:

> - `changelog_config`: should point out to the `git-cliff` configuration which we created in the previous section.

Reviewer: ```suggestion
- `changelog_config`: should point out to the `git-cliff` configuration we created in the previous section.
```

> - `dependencies_update`: setting this to `true` means that you want to run `cargo update` before the each release. Enabling this for updating the transitive dependencies wouldn't hurt since we will already be updating the main dependencies with Dependabot in the future steps.

Reviewer: ```suggestion
- `dependencies_update`: setting this to `true` means that you want to run `cargo update` before each release. Enabling this for updating the transitive dependencies wouldn't hurt since we will already be updating the main dependencies with Dependabot in the future steps.
```

- `git_tag_enable`: set this to `true` to create a tag for the new releases.
- `git_release_enable`: make sure to set this to `false` since we don't want `release-plz` to create a GitHub release for us since that part will be handled by `cargo-dist`.

As for the next step, we should create the _actual_ release automation that is responsible for running `release-plz` for each commit pushed to the repository and creating a "release PR".

**Q**: A release PR?

**A**: Release PR is a pull request that represents the next release and contains the necessary changes such as version bumps and changelog update.

<center>

![release PR](/release-pr.png)

[https://github.com/orhun/daktilo/pull/51](https://github.com/orhun/daktilo/pull/51)

</center>

> After the release PR is merged, `release-plz` detects that there is a version change in `Cargo.toml` and runs `cargo publish` to release the new version on crates.io

Reviewer: ```suggestion
After you merge the release PR, `release-plz` detects that there is a version change in `Cargo.toml` and runs `cargo publish` to release the new version on [https://crates.io/](crates.io).
```

Reviewer: Specifying the subject ("you" or "the maintainer" in this case), is useful for the reader to understand that merging the release pr is intentional, not part of an automation.

**Q**: Ah, so we probably need to set up secrets for the publish token etc., right?

> There is actually a bit more that we need to set up. You can check out [`release-plz` documentation](https://release-plz.ieni.dev/docs/github) about more detailed information but here is a quick summary:

Reviewer: ```suggestion
Yes! And a few permissions, too. You can check out [`release-plz` documentation](https://release-plz.ieni.dev/docs/github) about more detailed information but here is a quick summary:
```
less scary :P

- We are installing Rust stable ('cause why not?!)
- As the last step, we are running `release-plz`, heck yea!

**Q**: Whoa whoa whoa hey hey hey, what's up with `GITHUB_TOKEN: ${{ secrets.RELEASE_PLZ_TOKEN }}`? Just use `${{ secrets.GITHUB_TOKEN }}` bro, which is defined as default for each repository.

**A**: Good question! If you use `GITHUB_TOKEN` then it means that you are going to authenticate on behalf of GitHub Actions.

**Q**: So, what's wrong with that?

**A**: Well, workflows that are triggered by `on: push: tags` _won't run_ if the tag is created by GitHub Actions. So we are using a personal access token which is a way of saying "dear GitHub Actions, I am running `release-plz` and it is ME who is creating a tag so please run the further workflows". You can read more about it [here](https://release-plz.ieni.dev/docs/github/trigger).

Now we are simply ready to work on our project as usual and `release-plz` will create a release PR for our changes and we can merge it to create a new release when we are ready!

**Q**: What if I want to run/test `release-plz` locally though?

> **A**: You can if you want! The release PR is just the output of `release-plz update` command so you can run it locally and see what is changed in terms of changelog etc. Also, if you want to create a PR from the command-line you can simply run `release-plz release`!

Reviewer: ```suggestion
**A**: You can if you want! The release PR is just the output of `release-plz update` command so you can run it locally and see what is changed in terms of changelog etc. Also, if you want to create a PR from the command-line you can simply run `release-plz release-pr`!
```

---

### **cargo-dist**

> [cargo-dist](https://opensource.axo.dev/cargo-dist/) makes distributing binaries easier by breaking it down to multiple steps such as plan, build, publish and announce.

It is simply a tool for creating this beautiful release page with along with binary artifacts and installers:

<center>

![cargo-dist release](/cargo-dist-release.png)

[https://github.com/orhun/daktilo/releases/tag/v0.4.0](https://github.com/orhun/daktilo/releases/tag/v0.4.0)

</center>

Similar to `release-plz`, `cargo-dist` is a tool that we are not going to use locally but run from the CI instead. However, it is required to be installed to set up the CI (for itself) for the first time.

So let's start by installing `cargo-dist`:

```sh

> cargo install cargo-dist

Reviewer: ```suggestion
cargo install cargo-dist --locked
```
more secure!

```

> You can also use `cargo-dist` to install `cargo-dist` (i.e. the binary installer) if you wish:

Reviewer: ```suggestion
You can also use `cargo-dist-installer` to install `cargo-dist` (i.e. the binary installer) if you wish:
```
This recursion reads weird :P

```sh
cargo_dist_version="0.3.1"

curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v{version}/cargo-dist-installer.sh | sh
```

> We will also be able to install our application as shown above after `cargo dist` does its magic.

Reviewer: What do you mean with "install our application" here?

Author: Clarified: 920f8198954c4b6a361d3f25eb0517f5f2040ee6

    open-pull-requests-limit: 10

  # Maintain dependencies for GitHub Actions
  - package-ecosystem: github-actions
    directory: "/"
    schedule:
      interval: daily
    open-pull-requests-limit: 10
```

Now `Dependabot` will start creating pull requests for Rust/GitHub Actions dependencies when they become out-of-date:

<center>

![dependabot PR](/dependabot-pr.png)

[https://github.com/orhun/systeroid/pull/141](https://github.com/orhun/systeroid/pull/141)

</center>

---

### **Mergify**

> [`Mergify`](https://mergify.com/) is a CI/CD pipeline optimizer that manages your pull request, automates your GitHub workflow, and optimizes your CI costs.

We will be using `Mergify` to:

1. Automatically merge `Dependabot` pull requests.
2. Update to the latest `main` branch for pull requests.

Let's create the configuration in `.github/mergify.yml`:

```yml
pull_request_rules:
  - name: Automatic merge for Dependabot pull requests
    conditions:
      - author=dependabot[bot]
    actions:
      merge:
        method: squash

  - name: Automatic update to the main branch for pull requests
    conditions:
      - -conflict # skip PRs with conflicts
      - -draft # skip GH draft PRs

>       - -author=dependabot[bot] # skip dependabot PRs

Reviewer: It would be nice to skip release-plz PRs as well here. Because release-plz force pushes to the PR.
Maybe you can do it by the branch of the PR? I.e. if it starts with `release-plz` (not `release-plz/` because it's going to change to `release-plz-` in the next release).

Author: I will wait until the release is available :D
